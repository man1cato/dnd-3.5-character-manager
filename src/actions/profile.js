import database from '../firebase/firebase'


//SET PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
});

//READ PROFILE FROM FIREBASE
export const startSetProfile = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val();
            dispatch(setProfile(id, profile));
        });
    };
};

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
});

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates));
        });
    };
};

//DELETE PROFILE FROM STORE
export const deleteProfile = () => ({
    type: 'DELETE_PROFILE'
})

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
