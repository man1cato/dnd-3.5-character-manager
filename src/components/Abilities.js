import React from 'react';

const Abilities = (props) => (
  <div className="grid grid--abilities">
    <div className="grid--abilities__ability">Ability</div>
    <div className="grid--abilities__score">Score</div>
    <div className="grid--abilities__mod">Mod</div>
    <div className="grid--abilities__tempScore">Temp Score</div>
    <div className="grid--abilities__tempMod">Temp Mod</div>
    {Object.entries(props.abilities).map((ability, i) =>
      <h4 className="grid--abilities__ability" key={i}>{ability[1].name}</h4>
    )}
    {Object.entries(props.abilities).map((ability, i) =>
      <div className="grid--abilities__score" key={i}>{ability[1].score}</div>
    )}
    {Object.entries(props.abilities).map((ability, i) =>
      <div className="grid--abilities__mod" key={i}>{ability[1].mod}</div>
    )}
    {Object.entries(props.abilities).map((ability, i) =>
      <input
      className="grid--abilities__tempScore"
      key={i}
      id={ability[0]}
      type="text"
      value={ability[1].tempScore}
      onChange={props.onTempScoreChange}
      />
    )}
    {Object.entries(props.abilities).map((ability, i) =>
      <div className="grid--abilities__tempMod" key={i}>{ability[1].tempMod}</div>
    )}
  </div>
)

export default Abilities;
