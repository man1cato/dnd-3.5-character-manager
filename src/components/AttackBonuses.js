import React, { useState, useEffect, Fragment } from 'react'
import _ from 'lodash'
import update from 'immutability-helper'

import { calcBaseMeleeBonus, calcBaseRangedBonus, calcBaseGrappleBonus, convertInputValue } from '../utils/utils'


const AttackBonuses = ({ profileAttackBonusMods, abilities, baseAttackBonuses, size, handleUpdate }) => {  
    const attackBonusBases = {
        melee: _.map(baseAttackBonuses, (bab) => calcBaseMeleeBonus(bab, abilities.str.score, size)),
        ranged: _.map(baseAttackBonuses, (bab) => calcBaseRangedBonus(bab, abilities.dex.score, size)),
        grapple: _.map(baseAttackBonuses, (bab) => calcBaseGrappleBonus(bab, abilities.str.score, size))
    }
    const [attackBonusMods, setAttackBonusMods] = useState(_.mapValues(profileAttackBonusMods, (val) => val || '')) 
    const [attackBonusTotals, setAttackBonusTotals] = useState(
        _.mapValues(attackBonusMods, (mod, key) => _.map(attackBonusBases[key], (val) => Number(mod) + val ))
    ) 
    
    useEffect(() => {
        setAttackBonusTotals(_.mapValues(attackBonusMods, (mod, key) => _.map(attackBonusBases[key], (val) => Number(mod) + val)))
        handleUpdate({ attackBonusMods })
    }, [attackBonusMods])

    return (    
        <div className="grid--6col">
            <h5 className="grid__col1">Type</h5>
            <h5>Base</h5>
            <div></div>
            <h5>Mod</h5>
            <div></div>
            <h5>Total</h5>

            {_.map(['melee', 'ranged', 'grapple'], (key) => (
                <Fragment key={key}>
                    <div className="grid__col1">{_.startCase(key)}</div>
                    <div>{attackBonusBases[key].join(' / ')}</div>
                    <div>+</div>
                    <input
                        type="number"
                        id={key}
                        value={attackBonusMods[key]}
                        onChange={(e) => {
                            setAttackBonusMods(
                                update(attackBonusMods, {
                                    [key]: { $set: convertInputValue(e.target.value) }
                                })
                            )
                        }}
                    />
                    <div>=</div>
                    <div>{attackBonusTotals[key].join(' / ')}</div>
                </Fragment>
            ))}

        </div>
    )
}


export default AttackBonuses