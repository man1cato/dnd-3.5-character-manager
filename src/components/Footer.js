import React from 'react';
import {Link} from 'react-router-dom';
import SVGInline from "react-svg-inline";

import profileSVG from "../../public/images/profile.svg";
import statsSVG from "../../public/images/stats.svg";
import combatSVG from "../../public/images/combat.svg";
import equipmentSVG from "../../public/images/equipment.svg";
import spellsSVG from "../../public/images/spells.svg";
import companionSVG from "../../public/images/companion.svg";

const svgs = {
  profile: profileSVG,
  stats: statsSVG,
  combat: combatSVG,
  equipment: equipmentSVG,
  spells: spellsSVG,
  companion: companionSVG
}

const iconWrapper = (page) => (
  <Link to={`/${page}`} className="grid--footer__icon" key={page}>
    <SVGInline svg={svgs[page]} cleanup={true} height="100%" width="100%" fill="white" />
  </Link>
)

const Footer = () => (
  <footer className="container container--footer grid grid--footer">
    {Object.keys(svgs).map((svg) => iconWrapper(svg))}    
  </footer>
)

export default Footer;