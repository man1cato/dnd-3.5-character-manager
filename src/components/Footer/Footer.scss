@import '../../styles/base/settings';


.footer {
    position: fixed;
    bottom: 0px;
    height: 10vh;
    min-height: 60px;
    width: -webkit-fill-available;
    background: white;
    box-shadow: 0 0 5px $light-grey;
    padding: $m-size 0;
}

.footer--navbar {
    @extend .footer;
    background: #364051;
}

.footer--navbar__content {
    display: flex;
    justify-content: space-between;
    align-items: center;
    fill: white;
    > * {
        max-height: 40px;
        height: 100%;
        padding: 0 $s-size;
    }
    :first-child {
        padding-left: 0;
    }
    :last-child {
        padding-right: 0;
    }
}


