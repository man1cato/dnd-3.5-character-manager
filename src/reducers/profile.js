export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_PROFILE':
            return {
                id: action.id,
                ...action.fields
            };
        case 'EDIT_PROFILE':
            return {
                ...state,
                ...action.updates
            };
        default:
            return state;
    }
};
