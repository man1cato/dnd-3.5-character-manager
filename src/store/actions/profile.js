import database from '../../firebase/firebase'
import { startGetProfiles } from './profiles'

//SET SELECTED PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
})

//READ PROFILE FROM FIREBASE
export const startSetProfile = (uid, id) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val()
            dispatch(setProfile(id, profile))
        })
    }
}

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
            dispatch(startGetProfiles(uid))
            localStorage.setItem('selectedCharacterId', ref.key)
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
})

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates))
        })
    }
}

//REMOVE PROFILE FROM STORE
export const removeProfile = () => ({
    type: 'REMOVE_PROFILE'
})

//REMOVE PROFILE FROM FIREBASE
// export const startRemoveProfile = () => {
//     return (dispatch) => {
//         return database.ref(`users/${uid}/profiles/${id}`).remove().then(() => {
//             dispatch(removeProfile())
//         })
//     }
// }

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
