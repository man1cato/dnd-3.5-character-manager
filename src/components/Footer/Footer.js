import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import SVGInline from 'react-svg-inline'

import profileSVG from "../../../public/images/profile.svg"
import statsSVG from "../../../public/images/stats.svg"
import combatSVG from "../../../public/images/combat.svg"
import equipmentSVG from "../../../public/images/equipment.svg"
import spellbookSVG from "../../../public/images/spellbook.svg"
import companionSVG from "../../../public/images/companion.svg"
import './Footer.scss'


const FooterIcon = ({svgFile}) => (
	<SVGInline svg={svgFile} cleanup={true} height="100%" width="100%" fill="white" />
)

export const Footer = ({profile}) => (
	<footer className="footer--navbar">
		<div className="container footer--navbar__content" >
			<Link to="/profile" >
				<FooterIcon svgFile={profileSVG} />
			</Link>

			<Link to="/stats" >
				<FooterIcon svgFile={statsSVG} />
			</Link>

			<Link to="/combat" >
				<FooterIcon svgFile={combatSVG} />
			</Link>

			<Link to="/equipment" >
				<FooterIcon svgFile={equipmentSVG} />
			</Link>
			
			{profile.spellbook && 
				<Link to="/spellbook" >
					<FooterIcon svgFile={spellbookSVG} />
				</Link>
			}

			{profile.companion && 
				<Link to="/companion" >
					<FooterIcon svgFile={companionSVG} />
				</Link>
			}
		</div>
	</footer>
)

const mapStateToProps = (state) => ({
	profile: state.profile
})

export default connect(mapStateToProps)(Footer)