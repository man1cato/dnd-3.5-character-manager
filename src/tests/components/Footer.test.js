import React from 'react'
import {shallow} from 'enzyme'
import {Footer} from '../../components/Footer'
import profile from '../fixtures/profile';

const profileData = {
  id: profile.id,
  ...profile.fields
}

test('should render footer correctly', () => {
  const wrapper = shallow(<Footer profile={profileData}/>);
  expect(wrapper).toMatchSnapshot();
});
