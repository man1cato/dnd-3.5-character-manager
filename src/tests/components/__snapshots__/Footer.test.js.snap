// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render footer correctly 1`] = `
<footer
  className="container container--footer grid grid--footer"
>
  <Link
    replace={false}
    to="/profile"
  >
    <img
      className="grid--footer_img"
      src="images/person.svg"
    />
  </Link>
  <Link
    replace={false}
    to="/stats"
  >
    <img
      className="grid--footer_img"
      src="images/chart.svg"
    />
  </Link>
  <Link
    replace={false}
    to="/combat"
  >
    <img
      className="grid--footer_img"
      src="images/shield-with-swords.svg"
    />
  </Link>
  <Link
    replace={false}
    to="/equipment"
  >
    <img
      className="grid--footer_img"
      src="images/school-bag.svg"
    />
  </Link>
  <Link
    replace={false}
    to="/spells"
  >
    <img
      className="grid--footer_img"
      src="images/magic-wand.svg"
    />
  </Link>
  <Link
    replace={false}
    to="/companion"
  >
    <img
      className="grid--footer_img"
      src="images/paw-print.svg"
    />
  </Link>
</footer>
`;
